### PR TITLE
CA-204474: Wait for cancelled operations to finish before proceeding with hard shutdown/reboot

### PR DESCRIPTION
**Pull request by jonludlam**
Failing to do this has caused problems as it violates the principle that there should only be one operation per VM talking to xenopsd at a time.

### CODE REVIEW

File: ocaml/xapi/xapi_vm.ml [hunk 1]
```diff
@@ -254,9 +294,27 @@ let hard_shutdown ~__context ~vm =
...
+	| `Halted -> ()
+
+let hard_reboot ~__context ~vm =
+	cancel ~__context ~vm ~ops:[ `clean_shutdown; `clean_reboot; `pool_migrate; `call_plugin; `suspend ];
+	(* Cancelling operations can cause the VM to now be shutdown *)
+	begin
+	match Db.VM.get_power_state ~__context ~self:vm with
+	| `Running
+	| `Paused ->
+		Xapi_xenops.shutdown ~__context ~self:vm None;
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml]: I see that you are changing `reboot` into `shutdown` + `start`. Does that not have any consequences? I'm thinking of the semantic differences between shutdown+start and reboot?

File: ocaml/xapi/xapi_vm.ml [hunk 2]
```diff
@@ -236,14 +236,54 @@ let start_on  ~__context ~vm ~host ~start_paused ~force =
...
+      process from.Event_types.token
+    end else
+      ()
+  in
+  process ""
+
+let cancel ~__context ~vm ~ops =
+  let cancelled = List.filter_map (fun (task,op) ->
+    if List.mem op ops then begin
+      debug "Cancelling VM.%s for VM.hard_shutdown/reboot" (Record_util.vm_operation_to_string op);
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml]: Perhaps `info` rather than `debug`?

File: ocaml/xapi/xapi_vm.ml [hunk 3]
```diff
@@ -236,14 +236,54 @@ let start_on  ~__context ~vm ~host ~start_paused ~force =
...
+      Helpers.call_api_functions ~__context
+        (fun rpc session_id -> try Client.Task.cancel ~rpc ~session_id ~task:(Ref.of_string task) with _ -> ());
+      Some (Ref.of_string task)
+    end else None
+  ) (Db.VM.get_current_operations ~__context ~self:vm) in
+  wait_for_tasks ~__context ~tasks:cancelled
+
 
 let hard_shutdown ~__context ~vm =
+	cancel ~__context ~vm ~ops:[ `clean_shutdown; `clean_reboot; `hard_reboot; `pool_migrate; `call_plugin; `suspend ];
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml]: Do we need to worry about current operations on devices of the VM (VIFs, VBDs, ...)?

**Comment by robhoes** [comment]
"move" in the commit message of the second patch looks like "remove" if you look at that patch alone... ;)

File: ocaml/xapi/xapi_vm.ml (hunk 3 above)
Comment (jonludlam) [inline ocaml/xapi/xapi_vm.ml]: We don't currently, no. We probably should, but that's a far bigger change than we probably want at this stage.

File: ocaml/xapi/xapi_vm.ml [hunk 4]
```diff
@@ -254,9 +294,26 @@ let hard_shutdown ~__context ~vm =
...
+let hard_reboot ~__context ~vm =
+	cancel ~__context ~vm ~ops:[ `clean_shutdown; `clean_reboot; `pool_migrate; `call_plugin; `suspend ];
+	(* Cancelling operations can cause the VM to now be shutdown *)
+	begin
+	match Db.VM.get_power_state ~__context ~self:vm with
+	| `Running
+	| `Paused ->
+		Xapi_xenops.reboot ~__context ~self:vm None
+	| `Halted ->
+		start ~__context ~vm ~start_paused:false ~force:false
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml]: Calling `reboot` on a halted VM starts the VM, rather than raising `VM_BAD_POWER_STATE`? Isn't that an unnecessary change in behaviour?

Comment (jonludlam) [inline ocaml/xapi/xapi_vm.ml]: No, it was running when the call started (it got passed the allowed operations check), and very likely is only stopped because of the cancellation. This is the same behaviour as we have currently.

**Comment by robhoes** [comment]
Looks like all my questions were answered, so let's put this in now.
